xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,79 +240,243 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
+      throws ResourceInvalidException {
+    List<CidrRange> prefixRanges = new ArrayList<>();
+    List<CidrRange> sourcePrefixRanges = new ArrayList<>();
+    try {
+      for (io.envoyproxy.envoy.config.core.v3.CidrRange range : proto.getPrefixRangesList()) {
+        prefixRanges.add(new CidrRange(range.getAddressPrefix(), range.getPrefixLen().getValue()));
+      }
+      for (io.envoyproxy.envoy.config.core.v3.CidrRange range
+          : proto.getSourcePrefixRangesList()) {
+        sourcePrefixRanges.add(
+            new CidrRange(range.getAddressPrefix(), range.getPrefixLen().getValue()));
+      }
+    } catch (UnknownHostException e) {
+      throw new ResourceInvalidException("Failed to create CidrRange", e);
+    }
+    ConnectionSourceType sourceType;
+    switch (proto.getSourceType()) {
+      case ANY:
+        sourceType = ConnectionSourceType.ANY;
+        break;
+      case EXTERNAL:
+        sourceType = ConnectionSourceType.EXTERNAL;
+        break;
+      case SAME_IP_OR_LOOPBACK:
+        sourceType = ConnectionSourceType.SAME_IP_OR_LOOPBACK;
+        break;
+      default:
+        throw new ResourceInvalidException("Unknown source-type: " + proto.getSourceType());
+    }
+    return new FilterChainMatch(
+        proto.getDestinationPort().getValue(),
+        prefixRanges,
+        proto.getApplicationProtocolsList(),
+        sourcePrefixRanges,
+        sourceType,
+        proto.getSourcePortsList(),
+        proto.getServerNamesList(),
+        proto.getTransportProtocol());
+  }
 
+  @VisibleForTesting
+  static io.grpc.xds.HttpConnectionManager parseHttpConnectionManager(
+      HttpConnectionManager proto, boolean parseHttpFilter)
+      throws ResourceInvalidException {
     // Obtain max_stream_duration from Http Protocol Options.
     long maxStreamDuration = 0;
-    if (hcm.hasCommonHttpProtocolOptions()) {
-      HttpProtocolOptions options = hcm.getCommonHttpProtocolOptions();
+    if (proto.hasCommonHttpProtocolOptions()) {
+      HttpProtocolOptions options = proto.getCommonHttpProtocolOptions();
       if (options.hasMaxStreamDuration()) {
         maxStreamDuration = Durations.toNanos(options.getMaxStreamDuration());
       }
     }
 
-    // Parse filters.
-    List<NamedFilterConfig> filterChain = null;
-    if (parseFilter) {
-      filterChain = new ArrayList<>();
-      List<io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter>
-          httpFilters = hcm.getHttpFiltersList();
+    // Parse http filters.
+    List<NamedFilterConfig> filterConfigs = null;
+    if (parseHttpFilter) {
+      filterConfigs = new ArrayList<>();
+      Set<String> names = new HashSet<>();
       for (io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
-               httpFilter : httpFilters) {
+               httpFilter : proto.getHttpFiltersList()) {
         String filterName = httpFilter.getName();
+        if (!names.add(filterName)) {
```

Comment:
Thanks for adding this validation, which was missed in the original implementation.